File: src/main/java/com/afollestad/polarupgradetool/GradleMigrator.java
package com.afollestad.polarupgradetool;

import com.afollestad.polarupgradetool.jfx.UICallback;

import java.io.*;
import java.util.ArrayList;
import java.util.Collections;

/**
 * @author Aidan Follestad (afollestad)
 */
public class GradleMigrator {

    private final File mProject;
    private final File mLatest;
    private final ArrayList<String> mPropertyNames;
    private final ArrayList<String> mPropertyValues;
    private final UICallback uiCallback;

    public GradleMigrator(File project, File latest, String[] propertyNames, String[] propertyValues, UICallback uiCallback) {
        mProject = project;
        mLatest = latest;
        mPropertyNames = new ArrayList<>(propertyNames.length);
        Collections.addAll(mPropertyNames, propertyNames);
        mPropertyValues = new ArrayList<>(propertyValues.length);
        Collections.addAll(mPropertyValues, propertyValues);
        this.uiCallback = uiCallback;
    }

    @SuppressWarnings("ResultOfMethodCallIgnored")
    public boolean process() {
        final ArrayList<String> lines = new ArrayList<>();
        InputStream is = null;
        BufferedReader reader = null;

        try {
            is = new FileInputStream(mLatest);
            reader = new BufferedReader(new InputStreamReader(is));
            String line;

            while ((line = reader.readLine()) != null) {
                for (int i = 0; i < mPropertyNames.size(); i++) {
                    final String propertyName = mPropertyNames.get(i);
                    int start = line.indexOf(propertyName + " ");
                    if (start == -1) continue;
                    line = String.format("        %s %s", propertyName, mPropertyValues.get(i));
                }
                lines.add(line);
            }
        } catch (Exception e) {
            Main.LOG("[ERROR]: Failed to migrate a Gradle file: %s", e.getMessage());
            if (uiCallback != null) uiCallback.onErrorOccurred("Failed to migrate Gradle file:\n" + e.getMessage());
            e.printStackTrace();
            return false;
        } finally {
            Util.closeQuietely(reader);
            Util.closeQuietely(is);
        }

        mProject.delete();
        OutputStream os = null;
        BufferedWriter writer = null;

        try {
            os = new FileOutputStream(mProject);
            writer = new BufferedWriter(new OutputStreamWriter(os));

            for (int i = 0; i < lines.size(); i++) {
                if (i > 0) writer.newLine();
                writer.write(lines.get(i));
            }
        } catch (Exception e) {
            Main.LOG("[ERROR]: Failed to migrate a Gradle file: %s", e.getMessage());
            if (uiCallback != null) uiCallback.onErrorOccurred("Failed to migrate Gradle file:\n" + e.getMessage());
            e.printStackTrace();
            return false;
        } finally {
            Util.closeQuietely(writer);
            Util.closeQuietely(os);
        }

        Main.LOG("[INFO]: Migrated Gradle file %s", Main.cleanupPath(mProject.getAbsolutePath()));
        if (uiCallback != null)
            uiCallback.onStatusUpdate("Migrated Gradle file: " + Main.cleanupPath(mProject.getAbsolutePath()));
        return true;
    }
}

File: src/main/java/com/afollestad/polarupgradetool/GradleMigrator.kt
package com.afollestad.polarupgradetool

import com.afollestad.polarupgradetool.jfx.UICallback
import com.afollestad.polarupgradetool.utils.Util
import java.io.*
import java.nio.charset.Charset
import java.util.*

/**
 * @author Aidan Follestad (afollestad)
 */
class GradleMigrator(private val mProject: File, private val mLatest: File, private val uiCallback: UICallback?) {

    private fun processLineProperty(propertyName: String, line: String, propertyValue: String): String {
        val start = line.indexOf("$propertyName ")
        if (start == -1) return line
        return "        $propertyName $propertyValue"
    }

    @SuppressWarnings("ResultOfMethodCallIgnored")
    fun process(): Boolean {
        val lines = ArrayList<String>()
        try {
            mLatest.forEachLine(Charset.forName("UTF-8"), {
                var line = it
                line = line.replace("output.outputFile.parent, \"MyPolarPack",
                        "output.outputFile.parent, \"${Main.USER_APPNAME}")
                line = processLineProperty("applicationId", line, "\"${Main.USER_PACKAGE}\"")
                line = processLineProperty("versionName", line, "\"${Main.USER_VERSION_NAME}\"")
                line = processLineProperty("versionCode", line, Main.USER_VERSION_CODE)
                lines.add(line)
            })
        } catch (e: Exception) {
            MainBase.LOG("[ERROR]: Failed to migrate a Gradle file: ${e.message}")
            uiCallback?.onErrorOccurred("Failed to migrate Gradle file: ${e.message}")
            e.printStackTrace()
            return false
        }

        mProject.delete()
        var os: OutputStream? = null
        var writer: BufferedWriter? = null

        try {
            os = FileOutputStream(mProject)
            writer = BufferedWriter(OutputStreamWriter(os))

            for (i in lines.indices) {
                if (i > 0) writer.newLine()
                writer.write(lines[i])
            }
        } catch (e: Exception) {
            MainBase.LOG("[ERROR]: Failed to migrate a Gradle file: ${e.message}")
            uiCallback?.onErrorOccurred("Failed to migrate Gradle file: ${e.message}")
            e.printStackTrace()
            return false
        } finally {
            Util.closeQuietely(writer)
            Util.closeQuietely(os)
        }

        val cleanedPath = MainBase.cleanupPath(mProject.absolutePath)
        MainBase.LOG("[INFO]: Migrated Gradle file $cleanedPath")
        uiCallback?.onStatusUpdate("Migrated Gradle file: $cleanedPath")
        return true
    }
}